### PR TITLE
[Audit fixes] FOR-02: Inconsistent Deserialization of Address ID

### LINKED ISSUE #1134: FOR-02: Inconsistent Deserialisation of Address

### PR DESCRIPTION
**Pull request by cryptoquick**
**Summary of changes**
Changes introduced in this pull request:
- Compares serialized leb128 bytes to deserialized bytes.
- Adds a passing test case, in addition to 2 test cases for encodings meant to fail, as recommended in the audit document.
- These changes should not alter behavior in release build; it will only make the node behavior more consistent with that of the Lotus node.

**Reference issue to close (if applicable)**

Closes #1134 

**Other information and links**

- Explanation of changes: `from_leb_128` is too permissive of how leb128 bytes are deserialized, resulting in inconsistent behavior from how Lotus deserializes bytes. Lotus throws an error if there byte encodings are inconsistent.

### CODE REVIEW

File: vm/address/src/lib.rs [hunk 1]
```diff
@@ -303,7 +304,69 @@ pub(crate) fn from_leb_bytes(bz: &[u8]) -> Result<u64, Error> {
     let mut readable = bz;
 
     // write id to buffer in leb128 format
-    Ok(leb128::read::unsigned(&mut readable)?)
+    let id = leb128::read::unsigned(&mut readable)?;
+
+    if to_leb_bytes(id)? == bz {
+        Ok(id)
+    } else {
+        error!(
```
Comment (ec2) [inline vm/address/src/lib.rs:312]: Probably makes sense to have this be part of the InvalidAddressIDPayload variant instead of printing it out

**Review by creativcoder — APPROVED** [review]
:+1:

File: vm/address/src/lib.rs (hunk 1 above)
Comment (cryptoquick) [reply vm/address/src/lib.rs:312]: Like, maybe using a tuple in the enum variant to capture the id?

Comment (cryptoquick) [reply vm/address/src/lib.rs]: Take a look at these changes. Is this what you meant?
https://github.com/ChainSafe/forest/pull/1149/commits/2e5f084c5628bff42be67441aae92d563065fcb3

File: vm/address/Cargo.toml [hunk 2]
```diff
@@ -13,6 +13,7 @@ num-derive = "0.3.0"
 data-encoding = "2.1.2"
 data-encoding-macro = "0.1.7"
 leb128 = "0.2.1"
+log = "0.4.8"
```
Comment (ec2) [inline vm/address/Cargo.toml:16]: Think we can remove this now.

**Review by ec2 — APPROVED** [review]

**Pull request merged**